HW2/P2/parallel_vector.pyx

```diff
@@ -79,12 +79,32 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+   for r in range(repeat):
+       for idx in prange(src.shape[0], nogil=True, schedule=dynamic, num_threads=4):
+           #  Corner case.  This doesn't do anything to the counts, but
+           #  will cause a double lock.
+           if src[idx] == dest[idx] :
+               continue
+
+           # Prevent deadlock by acquiring both potential locks
+           # in order of index.
+           if src[idx] < dest[idx] :
```

Comment:
what if counts[src[idx]] changes between now and acquiring the lock? You usually want to check the same condition again after acquiring the lock